R/particle_filter.R

```diff
@@ -473,3 +500,73 @@ check_save_restart <- function(save_restart, data) {
 
   data$step_end[i]
 }
+
+
+history_single <- function(history_value, history_order, history_index,
+                           index_particle) {
+
+  ny <- nrow(history_value)
+
+  if (is.null(index_particle)) {
+    index_particle <- seq_len(ncol(history_value))
+  }
+
+  np <- length(index_particle)
+
+  nt <- ncol(history_order)
+
+  idx <- matrix(NA_integer_, np, nt)
+  for (i in rev(seq_len(ncol(idx)))) {
+    index_particle <- idx[, i] <- history_order[index_particle, i]
+  }
+
+  cidx <- cbind(seq_len(ny),
+                rep(idx, each = ny),
+                rep(seq_len(nt), each = ny * np))
+  ret <- array(history_value[cidx], c(ny, np, nt))
+  rownames(ret) <- names(history_index)
+  ret
+}
+
+history_nested <- function(history_value, history_order, history_index,
```

Comment:
Thanks for breaking this out. I implemented this a little while ago in dust (https://github.com/mrc-ide/dust/blob/ed782f0e9ace0422cc88fd2bd8dc7dc843183947/inst/include/dust/dust.hpp#L127-L165) and given how nasty it is we might centralise in a future PR so that this logic only exists in one place. This will be much easier to do with your change here